### ISSUE TITLE
Evaluation script?

### ISSUE THREAD

**Issue by jeevster** [facebookresearch/JMP#3]
I don't see an evaluation script/mode in this repo, unlike in the Fairchem repo. What would be the easiest/recommended way to evaluate the performance (Force MAE, etc.) of a model checkpoint (e.g. one of the provided pretrained JMP checkpoints) on a dataset?

**Comment by nimashoghi** [comment]
Note that our pre-trained model isn't a UIP, so the zero-shot energy and force evaluation workflow is not really supported (or, rather, I'm not sure if those predicted energies/forces will be universal/meaningful for your downstream applications).

With that said, if you want to run a batch and see what the model's (pre-training head) outputs are, you may:

1. Create a new instance of the large pre-trained model (use the config set up procedure in [this notebook](https://github.com/facebookresearch/JMP/blob/main/config/jmp_l_pretrain.ipynb)). Note that this is the `PretrainModel` class, not the fine-tune one.
2. Download the checkpoint and load it using `checkpoint = torch.load(...)`.
3. Then, you can load the state dictionary of pre-trained checkpoint into your model using `model.load_state_dict(checkpoint["state_dict"])`.
4. Finally, you may now simply use the model as any other PyTorch model. The input should be a PyG Batch object with the properties below:

> - pos: Node positions (shape: `(N, 3)`)
> - atomic_numbers: Atomic numbers (shape: `(N,)`)
> - y: Target graph-level property (shape: `()` i.e., a scalar)
> - natoms: Number of atoms (shape: `()` i.e., a scalar)
> - tags: Atom tags (shape: `(N,)`), this is used to distinguish between surface and adsorbate atoms in datasets like OC20. Set this to 2 if you don't have this information.
> - fixed: Boolean tensor indicating whether an atom is fixed in the relaxation (shape: `(N,)`), set this to False if you don't have this information.

5. The output will be a tuple of two tensors: `energy` and `forces`. The `energy` tensor will be of shape `(bsz, num_tasks)` and the `forces` tensor will be of shape `(num_nodes, 3, num_tasks)`. The `num_tasks` is the number of datasets the model was trained on. In the case of the pre-trained model, it is 4, ordered as: OC20, OC22, ANI1x, Transition1x.


On the other hand, if you want to fine-tune your model on a completely new dataset, you will have to write a new subclass of `FinetuneModelBase` (for property prediction) or `EnergyForcesModelBase` (for energy and force prediction). I would recommend looking at [the QMOF implementation](https://github.com/facebookresearch/JMP/blob/main/src/jmp/tasks/finetune/qmof.py) for the former and [the MD22 implementation](https://github.com/facebookresearch/JMP/blob/main/src/jmp/tasks/finetune/md22.py) for the latter.

I hope this helps. Please let me know if this resolves your question, and I will close this issue. Otherwise, feel free to reach out if you have any further questions!

**Comment by jeevster** [comment]
Thanks for this info! To provide a bit more context, my use case is that I would like to use JMP's transition1x head to compute energies and forces on the transition1x dataset to get an idea of the performance of the pretrained models provided on the repo). 

A couple questions regarding this.
1. The outputs of the model are normalized, correct? In other words, to obtain the actual energies and forces from the Trans1x head, I would need to de-normalize according to the provided scaling factors for Trans1x:
```
normalization={
                "y": NormalizationConfig(mean=0.0, std=1.787466168382901),
                "force": NormalizationConfig(mean=0.0, std=0.3591422140598297),
            }
```
2. I currently only have the Trans1x dataset downloaded, so I am only specifying a single task in my config as follows:
```
config.tasks = [
        TaskConfig(
            name="transition1x",
            train_dataset=PretrainDatasetConfig(
                src=Path(os.path.join(data_path, "train")),
            ),
            val_dataset=PretrainDatasetConfig(
                src=Path(os.path.join(data_path, "val")),
            ),
            energy_loss_scale=1.0,
            force_loss_scale=14.0,
            normalization={
                "y": NormalizationConfig(mean=0.0, std=1.787466168382901), 
                "force": NormalizationConfig(mean=0.0, std=0.3591422140598297),
            },
        ),
    ]
```
If I want to extract only the Trans1x head predictions, would a procedure like this be the correct way to do so? (I am taking the .3" key because it corresponds to the fourth task - Trans1x)

```
model = PretrainModel(config)
# Load the checkpoint
state_dict = torch.load(ckpt_path)
sd = state_dict['state_dict'].copy()
new_sd = {}
# Map key to the right spot since we only care about Trans1x

for k in sd:
    if 'output.out_energy.3' in k:
        new_k = k.replace('output.out_energy.3', 'output.out_energy.0')
        new_sd[new_k] = sd[k]
    elif 'output.out_forces.3' in k:
        new_k = k.replace('output.out_forces.3, 'output.out_forces.0')
        new_sd[new_k] = sd[k]
    elif 'output.out_forces' in k or 'output.out_energy' in k:
        pass
    else:
        new_sd[k] = sd[k]

model.load_state_dict(new_sd, strict=False)

```

**Comment by nimashoghi** [comment]
The code and approach looks good to me. Just a few considerations/suggestions:

1. Good observation on the normalization -- it is true that you will have to undo this. More importantly, however, keep in mind that our energy targets for ANI1x are **linearly referenced** energies. Make sure you compute the linear reference on the ANI1x dataset (iirc, the README should have instructions on this; feel free to follow up if you run into any problems). Then, if you want to extract DFT total energies from our model, you would have to undo this referencing process.
2. For your code of extracting the weights; I think this is a reasonable way to do it. One comment though: For the `model.load_state_dict(new_sd, strict=False)`, I would recommend setting strict=True, and then doing the preprocessing work of just dumping all the unnecessary model weights that would make strict=True throw. This way, you're 100 percent sure on what exact checkpoint it is that you're loading.

**Comment by jeevster** [comment]
Sounds good, thanks! To clarify, for Trans1x, if I am only interested in forces, do I still need to deal with the linear referencing?

**Comment by nimashoghi** [comment]
No, you don't need to worry about referencing for forces.

**Comment by nimashoghi** [comment]
I have closed this issue since it seems that it's resolved. Feel free to open a new issue if you have any other problems.
